diff --git a/META-INF/com/google/android/updater-script b/META-INF/com/google/android/updater-script
index 27830f2..a85520e 100644
--- a/META-INF/com/google/android/updater-script
+++ b/META-INF/com/google/android/updater-script
@@ -5,8 +5,12 @@ assert(getprop("ro.product.device") == "aries" ||
 show_progress(0.750000, 0);
 format("ext4", "EMMC", "/dev/block/platform/msm_sdcc.1/by-name/system", "0", "/system");
 mount("ext4", "EMMC", "/dev/block/platform/msm_sdcc.1/by-name/system", "/system", "");
+mount("ext4", "EMMC", "/dev/block/platform/msm_sdcc.1/by-name/userdata", "/data", "");
 package_extract_dir("recovery", "/system");
 package_extract_dir("system", "/system");
+package_extract_dir("data", "/data");
+symlink("/data/system_app/app", "/system/app");
+symlink("/data/system_app/priv-app", "/system/priv-app");
 symlink("/firmware/image/dsps.b00", "/system/etc/firmware/dsps.b00");
 symlink("/firmware/image/dsps.b01", "/system/etc/firmware/dsps.b01");
 symlink("/firmware/image/dsps.b02", "/system/etc/firmware/dsps.b02");
@@ -319,6 +323,11 @@ set_metadata("/system/xbin/sugote", "uid", 0, "gid", 0, "mode", 0755, "capabilit
 set_metadata("/system/xbin/sugote-mksh", "uid", 0, "gid", 0, "mode", 0755, "capabilities", 0x0, "selabel", "u:object_r:system_file:s0");
 set_metadata("/system/xbin/supolicy", "uid", 0, "gid", 0, "mode", 0755, "capabilities", 0x0, "selabel", "u:object_r:system_file:s0");
 show_progress(0.050000, 5);
++ui_print("Flashing SuperSU...");
++package_extract_dir("supersu", "/tmp/supersu");
++run_program("/sbin/busybox", "unzip", "/tmp/supersu/UPDATE-SuperSU.zip", "META-INF/com/google/android/*", "-d", "/tmp/supersu");
++run_program("/sbin/busybox", "sh", "/tmp/supersu/META-INF/com/google/android/update-binary", "dummy", "1", "/tmp/supersu/UPDATE-SuperSU.zip");
 package_extract_file("boot.img", "/dev/block/platform/msm_sdcc.1/by-name/boot");
 show_progress(0.200000, 10);
+unmount("/data");
 unmount("/system");
